presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java

```diff
@@ -1264,6 +1292,71 @@ public void testBadUrlMissingSchema()
         }
     }
 
+    @Test(timeOut = 10000)
+    public void testQueryCancellation()
+            throws Exception
+    {
+        try (Connection connection = createConnection("blackhole", "blackhole");
+                Statement statement = connection.createStatement()) {
+            statement.executeUpdate("CREATE TABLE test_cancellation (key BIGINT) " +
+                    "WITH (" +
+                    "   split_count = 1, " +
+                    "   pages_per_split = 1, " +
+                    "   rows_per_page = 1, " +
+                    "   page_processing_delay = '10s'" +
+                    ")");
+        }
+
+        CountDownLatch queryStarted = new CountDownLatch(1);
+        CountDownLatch queryFinished = new CountDownLatch(1);
+        AtomicReference<String> queryId = new AtomicReference<>();
+        AtomicReference<Throwable> queryFailure = new AtomicReference<>();
+
+        Future<?> queryFuture = executorService.submit(() -> {
+            try (Connection connection = createConnection("blackhole", "default");
+                    Statement statement = connection.createStatement();
+                    ResultSet resultSet = statement.executeQuery("SELECT * FROM test_cancellation")) {
+                queryId.set(resultSet.unwrap(PrestoResultSet.class).getQueryId());
+                queryStarted.countDown();
+                try {
+                    resultSet.next();
+                }
+                catch (SQLException t) {
+                    queryFailure.set(t);
+                }
+                finally {
+                    queryFinished.countDown();
+                }
+            }
+            return null;
+        });
+
+        queryStarted.await(10, SECONDS);
+        assertNotNull(queryId.get());
+        assertFalse(getQueryState(queryId.get()).isDone());
+        queryFuture.cancel(true);
+        queryFinished.await(10, SECONDS);
+        assertNotNull(queryFailure.get());
+        assertEquals(getQueryState(queryId.get()), FAILED);
+
+        try (Connection connection = createConnection("blackhole", "blackhole");
+                Statement statement = connection.createStatement()) {
+            statement.executeUpdate("DROP TABLE test_cancellation");
+        }
+    }
+
+    private QueryState getQueryState(String queryId)
```

Comment:
I used the `String` before, but than i must hardcode all the value which indicate that query is still running. From your former comment: 

```
This is brittle, since we might make changes to the state machine (and have recently, e.g. for transactions).

One fix is to use QueryState.valueOf(state).isDone(), but there's an easier way: select the end timestamp from the table. It will be null if the query is not done.
```